### ISSUE TITLE
[RFC] Add each X item discount reward option in promotion rules

### ISSUE THREAD

**Issue by IKarbowiak** [saleor/saleor#12912]
## Problem

Allow providing a discount on each X item on checkout and orders.

---
Blocked by: https://github.com/saleor/saleor/issues/12889
Should be also done after: https://github.com/saleor/saleor/issues/12911, but not required, otherwise the `reward_predicate` must be added here

## 1. General Assumptions

- Introduce a new`EACH_X_ITEM` reward type
- Introduce the `everyXItem` field that will keep the info on which item the discount should be applied
- Require providing the `rewardPredicate` (if not implemented yet, follow the steps from here this https://github.com/saleor/saleor/ 
- Require providing `rewardValueType`, `rewardValue` and `eachXItem`
- Update checkout and draft order calculations to apply the proper discount. Ensure the taxes are correctly calculated.

### Examples
Buy 2 shirts get 1 free
> 💡 `everyXItem` is required when `EACH_X_ITEM` reward type is chosen

```json
{
"promotionRule": {
  "rewardPredicate": {
    "__typename": "CategoryPredicate",
    "name": {
      "eq": "Shirt"
    }
  },
  "rewardType": "EACH_X_ITEM",
  "rewardValueType": "PERCENTAGE",
  "rewardValue": 100,
  "everyXItem": 3,
  }
}
```

### Acceptance Criteria

- [ ] The taxes should be correctly calculated

### Mutations
- Update the promotion inputs with `everyXItem` field
- Update the following mutations: `promotionCreate`, `PromotionRuleCreate`, `promotionRuleUpdate`
  - The `eachXItem` must be positive value above 0
  - When `EACH_X_ITEM` reward type is chosen require: `rewardPredicate`, `rewardValueType`, `rewardValue` and `eachXItem`

```graphql
input PromotionRuleInput @doc(category: "Discounts") {
  ...
  everyXItem: Int
}

input PromotionRuleCreateInput @doc(category: "Discounts") {
  ...
  everyXItem: Int
}

input PromotionRuleUpdateInput @doc(category: "Discounts") {
  ...
  everyXItem: Int
}
```

### Types
- Extend `PromotionRule` with `everyXItem `
- Extend `RewardTypeEnum` with `EACH_X_ITEM `

```graphql
type PromotionRule implements Node @doc(category: "Discounts"){
  ...
  everyXItem: Int
}

enum RewardTypeEnum @doc(category: "Discounts") {
  ...
  EACH_X_ITEM
}
```

## 3. Database changes

- Extend `PromotionRule` with  `every_x_item`
```python
class PromotionRule:
    ...
    every_x_item: IntField
```

## 4. To Do list

- [ ]  Extend the rewardType enum
- [ ]  Introduce the `eachXItem` field
- [ ]  Update the mutations with proper validation
- [ ]  Update the order and checkout price recalculations
- [ ]  Ensure tax for flat rates and Avatax are properly calculated

## 5. To Test

- [ ] Ensure the discount working correctly with fixed and percentage values
- [ ] Check with price ovverride

**Comment by IKarbowiak** [comment]
Postponed for now.
